x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,1335 @@
+
+---
+
+## Endpoints and utilities that will need to be adapted to the new schema
+
+### Utilities
+
+#### KQL filters and the `convertRulesFilterToKQL` method
+
+Across our application, both in the frontend and serverside, we use KQL filters to retrieve rules based on whether they are prebuilt rules or not - this means that the current behaviour of these values relies on the `immutable` field being set to either `true` or `false`.
+
+As mentioned before, we need to assume that at any point in time, there will be a mixture of rules whose saved object has already been migrated on Elasticsearch and others will not. This means that the retrieval of rules will need to maintain backwards compatibility: in order to determine if a rule is prebuilt, preferentially search for the existence of the `external_source` field; if that doesn't exist,  we should fall back to the legacy logic of checking a rule's `immutable` value.
+
+This means that we will need to update the constants and KQL filters that we have hardcoded in our application to reflect the new schema:
+
+_See source of rule params keys: [x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_fields.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_fields.ts)_
+
+Will need to update the `x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts` file, where the `convertRulesFilterToKQL` method is defined. This method is used both in the frontend and in the serverside, and translates rule filter options to KQL filters that Elasticsearch can understand.
+
+Here, we need to update the KQL filters and the logic for fetching Elastic prebuilt and custom rules, relying on `rule_source_type` but with fallback to `immutable`:
+
+_Source for `convertRulesFilterToKQL`: [x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/common/detection_engine/rule_management/rule_filtering.ts)_
+
+In order to retrieve Elastic prebuilt rules, we can filter for rules in which `alert.attributes.params.rule_source_type` is `external`, with a fallback to rules in which `alert.attributes.params.immutable` is `true`, for rules which have not had their schema migrated and `rule_source_type` does not still exists for them.
+
+> In the future, we can further distinguish between `external` rules by their source origin. If we introduce a new field within `external_source` that defines their origin (such as `repo_name`, `repo_id`, or similar), we can further filter by Elastic prebuilt rules, or rules that are externally sourced from a repository handled by the user, etc.
```

Comment:
cc @banderror @marshallmain 

After some discussion with @banderror, we decided to introduce a top-level `rule_source_type` field which can take only the values `external` and `internal`. This way we can filter Prebuilt rules from custom rules without having to rely on a subfield of the optional `external_source` field.

For the scope of Prebuilt Rule Customization, knowing the value of `rule_source_type` will be enough to know if a rule is Elastic prebuilt or custom.

In the future, thinking about DaC, a rule might be `custom` but `external`ly sourced, so here I described how we can possibly handle that in the future. I think at that point we will be rethinking how we display our filters in the application anyways: probably not simply filtering prebuilt vs custom, but more of a multivalue filter by the rule's origin. But I these decisions will need product input and are more part of the DaC epic anyways.

Reply:
We met on zoom to discuss this and I expressed my concern that `rule_source_type` introduces a relationship between sibling fields of the schema: `external_source` is only valid if `rule_source_type` is `external`, so we'd need to either enforce that constraint in additional checks everywhere we manage rules or encode that constraint in the schema by using `rule_source_type` as a discriminant. However, "additional checks everywhere" is bug prone because it's hard to ensure that this check is done in every place, and we already have one discriminant for rule params at the top level so adding another one is not easy.

Per our zoom discussion today, I'm now completely in favor of using the structure @banderror suggested with `type: 'internal' | 'external'` as a subfield of `ruleSource` and using `type` as the discriminant for the `ruleSource` object. It looks awkward at first with only `type` in `ruleSource` when `type` is `internal`, but I like how it makes the type very explicit and also makes it clear which rules have been migrated since every migrated rule should have `ruleSource`. This approach also avoids the problem with multiple discriminants for one object, since `type` is now a subfield of `ruleSource` so we have a top level `type` discriminant and a `ruleSource.type` discriminant that are independent.

I also briefly tested adding `ruleSource` to the API response schema and didn't see any problems with having a discriminated union for `ruleSource` within the broader rule discriminated union.

`x-pack/plugins/security_solution/common/api/detection_engine/model/rule_schema/rule_schemas.gen.ts`
```
export type ResponseFields = z.infer<typeof ResponseFields>;
export const ResponseFields = z.object({
  id: RuleObjectId,
  rule_id: RuleSignatureId,
  immutable: IsRuleImmutable,
  updated_at: z.string().datetime(),
  updated_by: z.string(),
  created_at: z.string().datetime(),
  created_by: z.string(),
  revision: z.number().int().min(0),
  related_integrations: RelatedIntegrationArray,
  required_fields: RequiredFieldArray,
  setup: SetupGuide,
  execution_summary: RuleExecutionSummary.optional(),
  rule_source: z.discriminatedUnion('type', [
    z.object({
      type: z.literal('external'),
      updated_at: z.string().datetime(),
    }),
    z.object({
      type: z.literal('internal')
    })
  ])
});
```